lib/axiom/fs/stream/generic_streams.js

```diff
@@ -0,0 +1,217 @@
+  this.paused_ = true;
+};
+
+/**
+ * @return {void}
+ */
+GenericStreams.prototype.resume = function() {
+  this.paused_ = false;
+  this.flushReadQueue_();
+};
+
+/**
+ * Consume one event from the stream. Return undefined if the
+ * stream is empty.
+ *
+ * @return {*}
+ */
+GenericStreams.prototype.read = function() {
+  if (!this.paused_) {
+    throw new AxiomError.Runtime('Cannot read: stream must be paused');
+  }
+  return this.read_();
+};
+
+/**
+ * Write data into the stream (for implementers of the stream only).
+ * Invoke the "onData" callback for all pending events if the callback is set.
+ *
+ * @param {!*} value
+ * @param {EventCallback=} opt_callback
+ * @return {void}
+ */
+GenericStreams.prototype.write = function(value, opt_callback) {
+  switch(this.getState_()) {
+    case GenericStreams.ConnectionState.CONNECTING:
+    case GenericStreams.ConnectionState.CONNECTED:
+      var item = value;
+      if (opt_callback) {
+        item = new EventWithCallback(value, opt_callback);
+      }
+      this.writeQueue_.enqueue(item);
+      this.flushWriteQueue_();
+      break;
+
+    case GenericStreams.ConnectionState.CLOSING:
+    case GenericStreams.ConnectionState.CLOSED:
+    default:
+      throw new AxiomError.Runtime('Cannot write: the connection is closed.');
+  }
+};
+
+/**
+ * @return {void}
+ */
+GenericStreams.prototype.flushReadQueue_ = function() {
+  while (!this.paused_) {
+    var item = this.read_();
+    if (!item)
+      break;
+
+    this.onData.fire(item);
+  }
+};
+
+/**
+ * @return {void}
+ */
+GenericStreams.prototype.flushWriteQueue_ = function() {
+  while(this.getState_() === GenericStreams.ConnectionState.CONNECTED) {
+    var item = this.writeQueue_.dequeue();
+    if (!item)
+      break;
+    this.handleSend_(item);
+  }
+};
+
+/**
+ * Consume one event from the stream. Return undefined if the
+ * stream is empty.
+ *
+ * @return {*}
+ */
+GenericStreams.prototype.read_ = function() {
+  return this.readQueue_.dequeue();
+};
+
+/**
+ * Connection states.
+ * @enum {number}
+ */
+ GenericStreams.ConnectionState = {
+  CONNECTING: 0,
+  CONNECTED: 1,
+  CLOSING: 2,
+  CLOSED: 3,
+};
+
+/**
+ * @protected
+ * @param {*} data
+ * @return {void}
+ */
+GenericStreams.prototype.handleReceive_ = function(data) {
+  this.readQueue_.enqueue(data);
+  if (this.paused_) {
+    this.onReadable.fire();
+  } else {
+    this.flushReadQueue_();
+  }
+};
+
+/**
+ * @protected
+ * @param {*} data
+ * @return {void}
+ */
+GenericStreams.prototype.handleSend_ = function(item) {
+};
+
+/**
+ * @protected
+ * @return {void}
+ */
+GenericStreams.prototype.handleConnected_ = function() {
+  this.flushWriteQueue_();
+}
+
+/**
+ * @protected
+ * @return {void}
+ */
+GenericStreams.prototype.handleClose_ = function() {
+  this.onClose.fire();
+}
+
+/**
+ * Return the state of the connection.
+ *
+ * @return {GenericStreams.ConnectionState}
+ * @protected
+ */
+GenericStreams.prototype.getState_ = function() {
```

Comment:
Cool!  I decided to add a bit more checking (check to see if the class has been overridden before throwing the error.  That way we can call the `super` javascript equivalent without it throwing.